distributed/cli/tests/test_dask_worker.py

```diff
@@ -682,3 +684,150 @@ def dask_setup(worker):
+            assert "closing nanny" in logs
+        else:
+            assert "nanny" not in logs
+        assert "stopping worker" in logs
+        assert "end worker" in logs
+        assert "timed out" not in logs
+        assert "signal" not in logs
+        assert "error" not in logs
+        assert "exception" not in logs
+        assert worker.returncode == 0
+    finally:
+        worker.kill()
+
+
+@pytest.mark.slow
+@pytest.mark.skipif(WINDOWS, reason="POSIX only")
+@pytest.mark.parametrize("nanny", ["--nanny", "--no-nanny"])
+@gen_cluster(client=True, nthreads=[])
+async def test_sigterm(c, s, nanny):
+    try:
+        kwargs = {}
+        worker = subprocess.Popen(
+            ["dask-worker", s.address, nanny],
+            stdout=subprocess.PIPE,
+            stderr=subprocess.STDOUT,
+        )
+        await c.wait_for_workers(1)
+
+        worker.send_signal(signal.SIGTERM)
+        stdout, stderr = worker.communicate()
+        logs = stdout.decode().lower()
+        assert stderr is None
+        assert f"signal {signal.SIGTERM}" in logs
+        if nanny == "--nanny":
+            assert "closing nanny" in logs
+        else:
+            assert "nanny" not in logs
+        assert "stopping worker" in logs
+        assert "end worker" in logs
+        assert "timed out" not in logs
+        assert "error" not in logs
+        assert "exception" not in logs
+        assert worker.returncode == 0
+    finally:
+        worker.kill()
+
+
+@pytest.mark.slow
+@pytest.mark.skipif(not WINDOWS, reason="Windows only")
+@pytest.mark.parametrize("nanny", ["--nanny", "--no-nanny"])
+@gen_cluster(client=True, nthreads=[])
+async def test_ctrl_c_event(c, s, nanny):
+    try:
+        worker = subprocess.Popen(
+            ["dask-worker", s.address, nanny],
+            stdout=subprocess.PIPE,
+            stderr=subprocess.STDOUT,
+            # Allow using CTRL_C_EVENT / CTRL_BREAK_EVENT
+            creationflags=subprocess.CREATE_NEW_PROCESS_GROUP,
+        )
+        await c.wait_for_workers(1)
+
+        worker.send_signal(signal.CTRL_C_EVENT)
+        stdout, stderr = worker.communicate()
+        logs = stdout.decode().lower()
+        assert stderr is None
+        if nanny == "--nanny":
+            assert "closing nanny" in logs
+        else:
+            assert "nanny" not in logs
+        assert "stopping worker" in logs
+        assert "end worker" in logs
+        assert "signal" not in logs
+        assert "timed out" not in logs
+        assert "error" not in logs
+        assert "exception" not in logs
+        assert worker.returncode == 0
+    finally:
+        worker.kill()
+
+
+@pytest.mark.slow
+@pytest.mark.skipif(not WINDOWS, reason="Windows only")
+@pytest.mark.parametrize("nanny", ["--nanny", "--no-nanny"])
+@gen_cluster(client=True, nthreads=[])
+async def test_ctrl_break_event(c, s, nanny):
+    try:
+        worker = subprocess.Popen(
+            ["dask-worker", s.address, nanny],
```

Comment:
Beware, a developer running tests may have a different `dask-worker` executable on their path.  This is actually pretty common because people install a released version of Dask, but then don't properly install the git version.  

To resolve this you could use `python -m distributed.cli.dask_worker` or you could use the `popen` context manager in utils_test.py

You might look at using popen regardless.  It has some advantages like printing out stdout/stderr.  It's also nice to be consistent across tests.

Reply:
When using `with popen`, I ran into an issue where my manual use of `.communicate()` on the yielded `Popen` object to obtain the contents of `stdout` and `stderr`  closed `stdout` and the `popen` context manager attempted to do the same on `__exit__`, running into a `ValueError: Read of a closed file`. If we want to use `with popen` consistently, one way to avoid manual use of `.communicate()` while allowing the user to obtain `stdout` and `stderr` might be to set `stdout` and `stderr` on the yielded object object here instead of writing to local variables that will be dropped after. 

https://github.com/dask/distributed/blob/198522bf767d7d8c1927df07951cd34a79109bd0/distributed/utils_test.py#L1247-L1252

Reply:
I think once these tests are working, we'll be able to see what the common parts of `utils_test.popen` can be refactored out